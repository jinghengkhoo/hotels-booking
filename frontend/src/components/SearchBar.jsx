import { useState, useEffect } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faSearch,
  faCalendarAlt,
  faUsers,
} from "@fortawesome/free-solid-svg-icons";
import CustomDatePicker from "./DatePicker";
import { useCombobox } from "downshift";
import axios from "axios";
import { useNavigate } from "react-router-dom";

const SearchBar = () => {
  const [destination, setDestination] = useState("");
  const [destinationId, setDestinationId] = useState("");
  const [suggestions, setSuggestions] = useState([]);
  const [startDate, setStartDate] = useState(new Date());
  const [endDate, setEndDate] = useState(new Date());
  const [guests, setGuests] = useState(1);
  const [rooms, setRooms] = useState(1);
  const navigate = useNavigate();

  useEffect(() => {
    if (destination.length > 0) {
      axios
        .get(`http://localhost:5555/api/destinations?query=${destination}`)
        .then((res) => {
          setSuggestions(res.data);
        })
        .catch((err) => console.error(err));
    } else {
      setSuggestions([]);
    }
  }, [destination]);

  const validateForm = () => {
    const newErrors = {};

    if (!destinationId) {
      newErrors.destinationId = "Please select a valid destination.";
    }
    if (!startDate) {
      newErrors.startDate = "Please select a start date.";
    }
    if (!endDate) {
      newErrors.endDate = "Please select an end date.";
    }
    if (guests <= 0) {
      newErrors.guests = "Please select the number of guests.";
    }
    if (rooms <= 0) {
      newErrors.rooms = "Please select the number of rooms.";
    }

    return Object.keys(newErrors).length === 0;
  };

  const handleFormSubmit = (e) => {
    e.preventDefault();
    if (validateForm()) {
      navigate("/hotels", {
        state: { destinationId, startDate, endDate, guests, rooms },
      });
    }
  };
  const {
    isOpen,
    getMenuProps,
    getInputProps,
    highlightedIndex,
    getItemProps,
  } = useCombobox({
    onInputValueChange: ({ inputValue }) => {
      setDestination(inputValue);
    },
    onSelectedItemChange: ({ selectedItem }) => {
      setDestination(selectedItem ? selectedItem.term : "");
      setDestinationId(selectedItem ? selectedItem.uid : "");
    },
    items: suggestions,
    itemToString: (item) => (item ? item.term : ""),
  });

  return (
    <div className="flex justify-center items-center py-8 font-montserrat">
      <form className="bg-primary p-4 rounded-lg shadow-md flex space-x-4" onSubmit={handleFormSubmit}>

        {/*Location snippet */}
        <div className="flex flex-col">
          <label htmlFor="destination" className="text-gray-700 mt-4 ml-4 font-bold">Location</label>
          <div className="flex items-center rounded-md p-1 pt-0">
            <input
              placeholder="where to next?"
              className="w-full px-3 py-2 bg-primary border-none focus:outline-none"
              {...getInputProps()}
              type="text"
              name="destination"
              id="destination"
            />
          </div>
          <ul
            className={`absolute w-56 bg-primary shadow-md max-h-80 overflow-scroll p-0 z-10 mt-1 ${!(isOpen && suggestions.length) && 'hidden'}`}
            {...getMenuProps()}
          >
            {isOpen &&
              suggestions.map((item, index) => (
                <li
                  key={item.id}
                  {...getItemProps({ item, index })}
                  className={`cursor-pointer px-4 py-2 ${highlightedIndex === index ? 'bg-blue-500 text-white' : 'bg-white text-black'}`}
                >
                  {item.term}
                </li>
              ))}
          </ul>
        </div>
        {/*Date snippet */}
        <div className="flex flex-col">
          <label htmlFor="destination" className="text-gray-700 mt-4 font-bold">Date Start</label>
          <div className="flex items-center py-1">
            <CustomDatePicker className="bg-primary" selectedDate={startDate} onChange={setStartDate} />
          </div>
        </div>
        <div className="flex items-center border border-gray-300 rounded-md p-2">
          <FontAwesomeIcon
            icon={faCalendarAlt}
            className="text-gray-500 mr-2"
          />
          <CustomDatePicker selectedDate={endDate} onChange={setEndDate} />
        </div>

        <div className="flex flex-col">
          <label htmlFor="destination" className="text-gray-700 mt-4 font-bold">Number of Travellers</label>
          <div className="flex items-center rounded-md">
            <select
              value={guests}
              onChange={(e) => setGuests(e.target.value)}
              className="w-full py-2 bg-primary focus:outline-none"
            >
              {[...Array(10).keys()].map((i) => (
                <option key={i + 1} value={i + 1}>
                  {i + 1} {i + 1 === 1 ? "Guest" : "Guests"}
                </option>
              ))}
            </select>
          </div>
        </div>

        <div className="flex flex-col">
          <label htmlFor="destination" className="text-gray-700 mt-4 font-bold">Number of Rooms</label>
          <div className="flex items-center rounded-md">
            <select
              value={rooms}
              onChange={(e) => setRooms(e.target.value)}
              className="w-full py-2 bg-primary focus:outline-none"
            >
              {[...Array(10).keys()].map((i) => (
                <option key={i + 1} value={i + 1}>
                  {i + 1} {i + 1 === 1 ? "Room" : "Rooms"}
                </option>
              ))}
            </select>
          </div>
        </div>
        <button type="submit" className="bg-neutral text-white px-16 m-2 rounded-md shadow-md hover:bg-blue-600">
          Search
        </button>
      </form>
    </div >
  );
};

export default SearchBar;