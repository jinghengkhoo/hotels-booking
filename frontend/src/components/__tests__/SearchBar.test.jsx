// src/components/__tests__/SearchBar.test.jsx
import React from 'react';
import { cleanup, render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import SearchBar from '../SearchBar';
import axios from 'axios';

jest.mock("axios");

describe('SearchBar', () => {
  const currency = 'USD';

  const mockCountries = [
    { term: 'United States', id: '1' },
    { term: 'Canada', id: '2' },
    { term: 'United Kingdom', id: '3' },
    { term: 'Australia', id: '4' },
    { term: 'Germany', id: '5' },
    { term: 'France', id: '6' },
    { term: 'Japan', id: '7' },
    { term: 'China', id: '8' },
    { term: 'India', id: '9' },
    { term: 'Brazil', id: '10' },
    { term: 'Singapore', id: '11' },
  ];

  afterEach(() => {
    cleanup();
    jest.clearAllMocks(); // Clear all mocks after each test
  });

  test('Test Nil: renders SearchBar component', async () => {
    axios.get.mockResolvedValueOnce({ data: mockCountries });
    await act(async () => {
      render(
        <Router>
          <SearchBar currency={currency} renderDatePicker={false} />
        </Router>
      );
    });

    screen.debug();

    //getting all elements by TestId
    const destination = screen.getByTestId('destination');
    const startDate = screen.getByTestId('startDate');
    const endDate = screen.getByTestId('endDate');
    const guests = screen.getByTestId('guests');
    const rooms = screen.getByTestId('rooms');
    const submit = screen.getByTestId('submit-button');

    expect(destination).toBeInTheDocument();
    expect(destination).toHaveAttribute('placeholder', 'where to next?');
    expect(startDate).toBeInTheDocument();
    expect(startDate).toHaveAttribute('placeholder', 'YYYY-MM-DD');
    expect(endDate).toBeInTheDocument();
    expect(endDate).toHaveAttribute('placeholder', 'YYYY-MM-DD');
    expect(guests).toBeInTheDocument();
    expect(rooms).toBeInTheDocument();
    expect(submit).toBeInTheDocument();
  });

  test('Test 0: destination input changes', async () => {
    axios.get.mockResolvedValueOnce({ data: mockCountries });
    await act(async () => {
      render(
        <Router>
          <SearchBar currency={currency} renderDatePicker={false} />
        </Router>
      );
    });

    const destination = screen.getByTestId('destination');
    expect(destination).toBeInTheDocument();
    expect(destination).toHaveAttribute('placeholder', 'where to next?');

    //changing the value of the destination
    fireEvent.change(destination, { target: { value: 'Singapore' } });
    expect(destination.value).toBe('Singapore');
  });

  test('Test 1: Should catch error and log to console on failed API call', async () => {
    const consoleErrorMock = jest.spyOn(console, 'error').mockImplementation(() => { });
    const mockError = new Error('API Error');

    axios.get.mockResolvedValueOnce({ data: mockCountries });
    axios.get.mockRejectedValueOnce(mockError);

    await act(async () => {
      render(
        <Router>
          <SearchBar currency={currency} renderDatePicker={false} />
        </Router>
      );
    });

    // Simulate user typing in the destination input
    const input = screen.getByPlaceholderText('where to next?');
    fireEvent.change(input, { target: { value: 'Test' } });

    // Wait for useEffect to complete and axios to throw an error
    await waitFor(() => {
      expect(consoleErrorMock).toHaveBeenCalledWith(mockError);
    });

    consoleErrorMock.mockRestore();
  });

  // // Test 2: Successful form submission
  // test('Successful form submission', async () => {
  //   axios.get.mockResolvedValueOnce({ data: mockCountries });
  //   await act(async () => {
  //     render(
  //       <Router>
  //         <SearchBar currency={currency} renderDatePicker={false} />
  //       </Router>
  //     );
  //   });

  //   // Fill out the form
  //   fireEvent.change(screen.getByTestId('destination'), { target: { value: 'Singapore' } });
  //   fireEvent.change(screen.getByTestId('startDate'), { target: { value: '2024-08-01' } });
  //   fireEvent.change(screen.getByTestId('endDate'), { target: { value: '2024-08-10' } });
  //   fireEvent.change(screen.getByTestId('guests'), { target: { value: '2' } });
  //   fireEvent.change(screen.getByTestId('rooms'), { target: { value: '1' } });

  //   console.log("destination", screen.getByTestId('destination').value);

  //   // Mock the destinationId (in a real app, this might be set by an API call)
  //   localStorage.setItem('destinationId', '123');

  //   // Submit the form
  //   console.log(screen.getByTestId('submit-button'));
  //   fireEvent.click(screen.getByTestId('submit-button'));
  //   // fireEvent.submit(screen.getByRole('button'));
  //   console.log("locsl storage dest", localStorage.getItem('destination'));
  //   console.log("locsl storage dest", localStorage.getItem('destinationId'));


  //   // Check local storage
  //   expect(localStorage.getItem('destinationId')).toBe('123');
  //   expect(localStorage.getItem('startDate')).toBe('2024-08-01');
  //   expect(localStorage.getItem('endDate')).toBe('2024-08-10');
  //   expect(localStorage.getItem('guests')).toBe('2');
  //   expect(localStorage.getItem('rooms')).toBe('1');

  //   // Check navigation
  //   expect(history.location.pathname).toBe('/hotels');
  //   expect(history.location.state).toEqual({
  //     destinationId: '123',
  //     startDate: '2024-08-01',
  //     endDate: '2024-08-10',
  //     guests: 2,
  //     rooms: 1,
  //     currency,
  //   });
  // });

  // Test 6: Test to check if the start / end date input changes
  test('Test 6: startDate input changes', async () => {
    axios.get.mockResolvedValueOnce({ data: mockCountries });
    await act(async () => {
      render(
        <Router>
          <SearchBar currency={currency} renderDatePicker={false} />
        </Router>
      );
    });

    const startDate = screen.getByTestId('startDate');
    expect(startDate).toBeInTheDocument();

    //changing the value of the startDate
    fireEvent.change(startDate, { target: { value: '2024-08-10' } });
    expect(startDate.value).toBe('2024-08-10');
  });

  test('Test 6: valid endDate change', async () => {
    axios.get.mockResolvedValueOnce({ data: mockCountries });
    await act(async () => {
      render(
        <Router>
          <SearchBar currency={currency} renderDatePicker={false} />
        </Router>
      );
    });
    const endDate = screen.getByTestId('endDate');
    expect(endDate).toBeInTheDocument();

    //changing the value of the endDate
    fireEvent.change(endDate, { target: { value: '2024-08-12' } });
    expect(endDate.value).toBe('2024-08-12');
  });

  test('Test 8: valid guest change', async () => {
    axios.get.mockResolvedValueOnce({ data: mockCountries });
    await act(async () => {
      render(
        <Router>
          <SearchBar currency={currency} renderDatePicker={false} />
        </Router>
      );
    });

    const guests = screen.getByTestId('guests');
    expect(guests).toBeInTheDocument();

    //changing the value of the guests
    fireEvent.change(guests, { target: { value: '2' } });
    expect(guests.value).toBe('2');
  });

  // Test 8: Test to check if the guests/rooms change
  test('Test 8: valid rooms change', async () => {
    axios.get.mockResolvedValueOnce({ data: mockCountries });
    await act(async () => {
      render(
        <Router>
          <SearchBar currency={currency} renderDatePicker={false} />
        </Router>
      );
    });

    const rooms = screen.getByTestId('rooms');
    expect(rooms).toBeInTheDocument();

    //changing the value of the rooms
    fireEvent.change(rooms, { target: { value: '2' } });
    expect(rooms.value).toBe('2');
  });

  test('matches snapshot', () => {
    const tree = renderer.create(
      <Router>
        <SearchBar currency={currency} renderDatePicker={false} />
      </Router>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
