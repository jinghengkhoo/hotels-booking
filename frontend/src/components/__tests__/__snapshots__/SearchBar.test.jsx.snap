// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar matches snapshot 1`] = `
<div>
  <div
    className="flex justify-center items-center py-8 font-montserrat"
  >
    <form
      className="bg-base-100 p-4 rounded-lg shadow-md flex space-x-16 flex-col md:flex-row space-y-4 md:space-y-0 md:space-x-4"
      onSubmit={[Function]}
    >
      <div
        className="flex flex-col"
      >
        <label
          className="text-gray-700 mt-4 ml-4 font-bold"
          id="destination"
        >
          Location
        </label>
        <div
          className="flex items-center rounded-md p-1 pt-0"
        >
          <input
            aria-activedescendant=""
            aria-autocomplete="list"
            aria-controls="downshift-:r0:-menu"
            aria-expanded={false}
            aria-labelledby="downshift-:r0:-label"
            autoComplete="off"
            className="w-full px-3 py-2 bg-base-100 mt-2 border-none focus:outline-none"
            data-testid="destination"
            id="destination"
            name="destination"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            placeholder="where to next?"
            role="combobox"
            type="text"
            value="Singapore"
          />
        </div>
        <ul
          aria-labelledby="downshift-:r0:-label"
          className="absolute w-56 bg-base-100 shadow-md max-h-80 overflow-scroll p-0 z-10 mt-28 hidden"
          id="downshift-:r0:-menu"
          onMouseLeave={[Function]}
          role="listbox"
        />
      </div>
      <div
        className="flex flex-col"
      >
        <label
          className="text-gray-700 mt-4 font-bold"
          id="startDate"
        >
          Date Start
        </label>
        <div
          className="bg-base-100 flex items-center mt-2 py-1"
        >
          <input
            className="form-input bg-base-100 flex items-center py-1"
            data-testid="startDate"
            onChange={[Function]}
            pattern="\\d{4}-\\d{2}-\\d{2}"
            placeholder="YYYY-MM-DD"
            required={true}
            type="text"
            value="2024-08-04"
          />
        </div>
      </div>
      <div
        className="flex flex-col"
      >
        <label
          className="text-gray-700 mt-4 font-bold"
          id="endDate"
        >
          Date End
        </label>
        <div
          className="bg-base-100 flex items-center mt-2 py-1"
        >
          <input
            className="form-input bg-base-100 flex items-center py-1"
            data-testid="endDate"
            onChange={[Function]}
            pattern="\\d{4}-\\d{2}-\\d{2}"
            placeholder="YYYY-MM-DD"
            required={true}
            type="text"
            value="2024-08-11"
          />
        </div>
      </div>
      <div
        className="flex flex-col"
      >
        <label
          className="text-gray-700 mt-4 font-bold"
          id="guests"
        >
          Number of Travelers
        </label>
        <div
          className="flex items-center rounded-md mt-2"
        >
          <select
            className="w-full py-2 bg-base-100 focus:outline-none"
            data-testid="guests"
            onChange={[Function]}
            value={1}
          >
            <option
              value={1}
            >
              1
               
              Guest
            </option>
            <option
              value={2}
            >
              2
               
              Guests
            </option>
            <option
              value={3}
            >
              3
               
              Guests
            </option>
            <option
              value={4}
            >
              4
               
              Guests
            </option>
            <option
              value={5}
            >
              5
               
              Guests
            </option>
            <option
              value={6}
            >
              6
               
              Guests
            </option>
            <option
              value={7}
            >
              7
               
              Guests
            </option>
            <option
              value={8}
            >
              8
               
              Guests
            </option>
            <option
              value={9}
            >
              9
               
              Guests
            </option>
            <option
              value={10}
            >
              10
               
              Guests
            </option>
            <option
              value={11}
            >
              11
               
              Guests
            </option>
            <option
              value={12}
            >
              12
               
              Guests
            </option>
            <option
              value={13}
            >
              13
               
              Guests
            </option>
            <option
              value={14}
            >
              14
               
              Guests
            </option>
            <option
              value={15}
            >
              15
               
              Guests
            </option>
            <option
              value={16}
            >
              16
               
              Guests
            </option>
            <option
              value={17}
            >
              17
               
              Guests
            </option>
            <option
              value={18}
            >
              18
               
              Guests
            </option>
            <option
              value={19}
            >
              19
               
              Guests
            </option>
            <option
              value={20}
            >
              20
               
              Guests
            </option>
            <option
              value={21}
            >
              21
               
              Guests
            </option>
            <option
              value={22}
            >
              22
               
              Guests
            </option>
            <option
              value={23}
            >
              23
               
              Guests
            </option>
            <option
              value={24}
            >
              24
               
              Guests
            </option>
            <option
              value={25}
            >
              25
               
              Guests
            </option>
          </select>
        </div>
      </div>
      <div
        className="flex flex-col"
      >
        <label
          className="text-gray-700 mt-4 font-bold"
          id="rooms"
        >
          Number of Rooms
        </label>
        <div
          className="flex mt-2 items-center rounded-md"
        >
          <select
            className="w-full py-2 bg-base-100 focus:outline-none"
            data-testid="rooms"
            onChange={[Function]}
            value={1}
          >
            <option
              value={1}
            >
              1
               
              Room
            </option>
            <option
              value={2}
            >
              2
               
              Rooms
            </option>
            <option
              value={3}
            >
              3
               
              Rooms
            </option>
            <option
              value={4}
            >
              4
               
              Rooms
            </option>
            <option
              value={5}
            >
              5
               
              Rooms
            </option>
            <option
              value={6}
            >
              6
               
              Rooms
            </option>
            <option
              value={7}
            >
              7
               
              Rooms
            </option>
            <option
              value={8}
            >
              8
               
              Rooms
            </option>
            <option
              value={9}
            >
              9
               
              Rooms
            </option>
            <option
              value={10}
            >
              10
               
              Rooms
            </option>
          </select>
        </div>
      </div>
      <div
        className="flex flex-col"
      >
        <button
          className="bg-accent text-white px-8 py-4 mt-6 rounded-md shadow-md hover:bg-accent"
          data-testid="submit-button"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <div
    className="flex justify-center"
  />
</div>
`;
